src/trace/implementations/mod.rs

```diff
@@ -239,16 +225,90 @@ impl<T: Columnation> RetainFrom<T> for TimelyStack<T> {
     }
 }
 
-/// Trait for types used as offsets into an ordered layer.
-/// This is usually `usize`, but `u32` can also be used in applications
-/// where huge batches do not occur to reduce metadata size.
-pub trait OrdOffset: Copy + PartialEq + Add<Output=Self> + Sub<Output=Self> + TryFrom<usize> + TryInto<usize>
-{}
+use std::convert::TryInto;
 
-impl<O> OrdOffset for O
-where
-    O: Copy + PartialEq + Add<Output=Self> + Sub<Output=Self> + TryFrom<usize> + TryInto<usize>,
-{}
+/// A list of unsigned integers that uses `u32` elements as long as they are small enough, and switches to `u64` once they are not.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Debug, Abomonation)]
+pub struct OffsetList {
+    /// Offsets that fit within a `u32`.
+    smol: Vec<u32>,
+    /// Offsets that either do not fit in a `u32`, or are inserted after some offset that did not fit.
+    chonk: Vec<u64>,
```

Comment:
What about extending this to `u16`? It would save some bytes for batches with less then 64k elements, which should be quite frequent.

Reply:
I'm worried that the added pain when you use things that don't fit in a `u32` feels more acute for things that don't fit in a `u16`. Maybe adding this makes sense but it's now shaving off just two bytes, and .. I'm up for it if we don't care about the CPU cost, but someone using DD probably does (more than a hypothetical 4B+ user).